### PR TITLE
tECDSA signing: TSS round 4

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Refs: https://github.com/keep-network/keep-core/issues/3042
Depends on: https://github.com/keep-network/keep-core/pull/3241

Here we implement the TSS round four for the tECDSA signing protocol. This round expects simple broadcast messages produced in round three as input and produces a simple broadcast message as output.

### Next steps

This PR is just a part of the entire work regarding the tECDSA signing test loop. Please refer to the description of https://github.com/keep-network/keep-core/issues/3042 for the full roadmap and the current status of the work. At this point, the next step will be introducing the fifth round of TSS signing.

### CODE REVIEW

File: pkg/tecdsa/signing/gen/pb/message.proto [hunk 1]
```diff
@@ -26,4 +26,10 @@ message TSSRoundThreeMessage {
     uint32 senderID = 1;
     bytes payload = 2;
     string sessionID = 3;
+}
+
+message TSSRoundFourMessage {
+    uint32 senderID = 1;
+    bytes payload = 2;
+    string sessionID = 3;
 }
```
Comment (tomaszslabon) [inline pkg/tecdsa/signing/gen/pb/message.proto:35]: I would consider adding a newline at the end of the file (it's a convention on unix-like systems).

File: pkg/tecdsa/signing/message.go [hunk 2]
```diff
@@ -91,3 +91,23 @@ func (trtm *tssRoundThreeMessage) SenderID() group.MemberIndex {
...
+	payload   []byte
+	sessionID string
+}
+
+// SenderID returns protocol-level identifier of the message sender.
+func (trfm *tssRoundFourMessage) SenderID() group.MemberIndex {
+	return trfm.senderID
+}
+
+// Type returns a string describing an tssRoundFourMessage type for
```
Comment (pdyraga) [inline pkg/tecdsa/signing/message.go:109]:
```suggestion
// Type returns a string describing a tssRoundFourMessage type for
```

File: pkg/tecdsa/signing/states.go [hunk 3]
```diff
@@ -373,9 +378,94 @@ func (trts *tssRoundThreeState) Next() (state.State, error) {
...
+	}, nil
 }
 
 func (trts *tssRoundThreeState) MemberIndex() group.MemberIndex {
 	return trts.member.id
 }
+
+// tssRoundFourState is the state during which members broadcast TSS
+// round four messages.
+// `tssRoundMessage`s are valid in this state.
```
Comment (pdyraga) [inline pkg/tecdsa/signing/states.go:395]:
```suggestion
// `tssRoundFourMessage`s are valid in this state.
```

File: pkg/tecdsa/signing/protocol_test.go [hunk 4]
```diff
@@ -1293,4 +1482,68 @@ func initializeTssRoundThreeMembersGroup(
...
+			tssRoundFourMembers,
+			member.initializeTssRoundFour(),
+		)
+		tssRoundThreeMessages = append(tssRoundThreeMessages, tssRoundThreeMessage)
+
+		cancelCtx()
+	}
+
+	return tssRoundFourMembers, tssRoundThreeMessages, nil
 }
```
Comment (tomaszslabon) [inline pkg/tecdsa/signing/protocol_test.go:1549]: It seems the files is not formatted with `go fmt`.

**Review by pdyraga — APPROVED** [review]

**Comment by pdyraga** [comment]
Left some non-blocking questions, let's address them in the last PR in the signing series or even a separate one.

File: pkg/tecdsa/signing/gen/pb/message.proto (hunk 1 above)
Comment (lukasz-zimnoch) [reply pkg/tecdsa/signing/gen/pb/message.proto:35]: Fixed after https://github.com/keep-network/keep-core/pull/3252 was merged to `main`.

File: pkg/tecdsa/signing/protocol_test.go (hunk 4 above)
Comment (lukasz-zimnoch) [reply pkg/tecdsa/signing/protocol_test.go:1549]: Fixed after https://github.com/keep-network/keep-core/pull/3252 was merged to main.

File: pkg/tecdsa/signing/gen/pb/message.proto [hunk 5]
```diff
@@ -26,4 +26,10 @@ message TSSRoundThreeMessage {
     uint32 senderID = 1;
     bytes payload = 2;
     string sessionID = 3;
+}
+
+message TSSRoundFourMessage {
+    uint32 senderID = 1;
+    bytes payload = 2;
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/tecdsa/signing/gen/pb/message.proto:33]: Addressed by https://github.com/keep-network/keep-core/pull/3271

File: pkg/tecdsa/signing/marshaling.go [hunk 6]
```diff
@@ -160,6 +160,34 @@ func (trtm *tssRoundThreeMessage) Unmarshal(bytes []byte) error {
...
+// network communication.
+func (trfm *tssRoundFourMessage) Marshal() ([]byte, error) {
+	return (&pb.TSSRoundFourMessage{
+		SenderID:  uint32(trfm.senderID),
+		Payload:   trfm.payload,
+		SessionID: trfm.sessionID,
+	}).Marshal()
+}
+
+// Unmarshal converts a byte array produced by Marshal to an tssRoundFourMessage.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/tecdsa/signing/marshaling.go:173]: Addressed by https://github.com/keep-network/keep-core/pull/3271

File: pkg/tecdsa/signing/message.go (hunk 2 above)
Comment (lukasz-zimnoch) [reply pkg/tecdsa/signing/message.go:109]: Addressed by https://github.com/keep-network/keep-core/pull/3271

File: pkg/tecdsa/signing/states.go (hunk 3 above)
Comment (lukasz-zimnoch) [reply pkg/tecdsa/signing/states.go:395]: Addressed within TSS round 5 PR: 50bbe745

**Pull request merged**